### ISSUE TITLE
compiler-cli ngfactory files are huge - make it more prone to mangle

### ISSUE THREAD

**Issue by sod** [angular/angular#10812]
**I'm submitting a ...**  (check one with "x")

```
[x] feature request
```

**Current behavior**

The *.ngfactory.ts files generated by `ngc` are huge. In my case 1.5MB in total (80 Components). And the code doesn't become much smaller with minify/mangle as there is `this.` before 95% of all statements.

Example Component: [gist.github.com/sod/049ce...](https://gist.github.com/sod/049ce1f29f6c6d83bcfdd99ecef49513)
`ngc` output for that component: [gist.github.com/sod/ca6dc...](https://gist.github.com/sod/ca6dc8aab403481ec84a7f308906a25a)

If I transcode & mangle this factory alone, the output is 66kb for this one component.

**Expected/desired behavior**

More minify friendly output of `ngc`.
- **Angular version:** 2.0.0-rc.5

**Comment by kylecordes** [comment]
The template precompiler described is amenable to shrinking the overall size of a minified/tree-shaken A2 program. It achieves this because:

* It does not mention types/files/function of A2 you didn't actually use.
* It avoids the need to include the compiler itself in a deployed program.

However, it appears from this example code above, that the compiler output itself is considerably larger than its input template.

I wonder if this might imply that, for projects above a certain size, the overall minified-production-tree-shaken results might be *larger* with the precompiler.

(Maybe-to-do for me: make angular2-stress-test generate more complex templates, so it could be used to test this hypotheses easily.)

**Comment by sod** [comment]
It is larger for me. 

* minify + mangle + JIT  (with @angular/compile) = 1.4 MB
* minify + mangle + AoT (without @angular/compile) = 2.2 MB

Both treeshaken via rollup and angular/.../esm/src files

**Comment by robwormald** [comment]
generated template files will, at a certain point, almost always end up larger than just the compiler - what stays consistent is the *performance* in such cases.

Startup times in the sampling I was doing this weekend:
```
Number of components / AoT startup time / JiT startup time
5 - 295ms / 435ms
50 - 322ms / 653ms
500 - 946ms / 2320ms
```

This is part of the reason why we like lazy loading with the router - you don't have to load megalithic bundles, and you get the performance boost from precompiled templates.

**Comment by kylecordes** [comment]
Rob - is it feasible to mix and match? This would mean precompiled templates in the initial bundle, for a smallest-possible app startup. Then JIT templates, slower to start but smaller, for later-loaded additional code.

**Comment by CaptainCodeman** [comment]
One of the great things about HTML is that it's a very terse, yet still readable, representation of the thing to be rendered vs most programming languages. Even with the ugly syntax that Angular 2 adds on top, the template string is always going to be way smaller than the javascript representation of it. So once you have more than even a small number of components you can easily be generating more code to represent the templates than what it would have taken to include the runtime compiler.

> generated template files will, at a certain point, almost always end up larger than just the compiler

Using the [webpack visualizer](https://chrisbateman.github.io/webpack-visualizer/) showed it only represented about 5% of my app bundle and that + the templates themselves is always likely to be quickly dwarfed by the code that the compiler is spitting out in a real, non hello-world, app. That's the most I'm ever going to save from compiling them and I distinctly remember the promise was that it would make bundles smaller, not just faster, definitely smaller (because the conversation was about how huge the bundles were).

What happened to all the promises of the offline compiler and bundle size? Splitting things up doesn't reduce the size. Yes, it allows lazy loading of _parts_ of the app but all that code still needs to be loaded and parsed eventually. You haven't eaten less pie just because you cut it into pieces first.

**Comment by TheLarkInn** [comment]
Not to mention thanks to string fragmentation, gzipping should allow for great string compression.

**Comment by kylecordes** [comment]
Interesting point from @CaptainCodeman about HTML being terse. I've always thought of it as tediously verbose. On various big Angular 1 and other pre-A2 projects, we have mostly used Jade as an alternate syntax, to tame the painful duplication, verbosity, and excessive syntax noise in HTML.

JSX is as or more concise than HTML. TSX is nearly so, and is also typed (please can we have this in A2 template?).

I think the more prominent issue here is that the template compiler (intentionally) optimized for speed over size. This has been discussed prominently, from the podium at ng-conf, where one of the core team (sorry, I forget who) showed that the compiler emits code optimized performance much "as if" a human has manually hand-unrolled each bit to create the shortest possible execution path for each kind of change.

This suggests a path for improvement - a way to ask the (pre-)compiler to generate slightly less fast, but more concise, code. Compilers for languages like C have had this for approximately my entire lifetime, this compiler for this HTML-like language could get it also.

**Comment by vicb** [comment]
@kylecordes the AoT compiler is quite "young". We have plan to improve on both code size and speed in the future. Stay tuned.

**Comment by tbosch** [comment]
See https://github.com/angular/angular/issues/11683

**Comment by sod** [comment]
Tested 2.2.1 because of generated size improvements mentioned in the changelog since 2.0.0

Our application ngfactory files shrank from 1.8 MB to 1.7 MB, but there are 34 new ngfactory files in node_modules/@angular/... that in sum result in a build exactly as big as before.

**Comment by aitboudad** [comment]
There is lot of unnecessary render empty text 
```
this._text_7 = this.renderer.createText(this._el_6,'\n                    ',null);
```
using html minify before parsing templates reduced my app size by ~200kb

```ts
parse(....) {
    const Minimize = require('minimize'), 
      minimize = new Minimize({ quotes: true, dom: { lowerCaseAttributeNames: false }});
    if (template) {
        template = minimize.parse(template);
    }
    if (templateUrl) {
        templateUrl = minimize.parse(templateUrl);
    }
    ....
}
```

**Comment by sod** [comment]
Closing my issue. I think it's fair to say that the filesize from a framework that renders the view inside the browser won't ever become small when we dump the entire application template code into the build.

Played around with [webpack 2.2](https://github.com/webpack/webpack) + [angular2-router-loader](https://github.com/brandonroberts/angular2-router-loader) + @angular/router to split the build into chunks and it works surprisingly good.

For anyone interested. This is the barebone `webpack.config.js` I started with:

```typescript
module.exports = {
    entry: {
        application: ['./app/bootstrap.ts'],
    },
    output: {
        path: __dirname + '/web',
        filename: 'js/[name].build.js',
        chunkFilename: 'js/[name].build.js',
    },
    resolve: {
        extensions: ['.ts', '.js']
    },
    module: {
        exprContextCritical: false,
        loaders: [{
            test: /\.ts$/,
            loaders: [
                'awesome-typescript-loader?configFileName=./app/bootstrap.ts',
                'angular-router-loader',
            ]
        }]
    }
};
```

And the `routes` for `RouterModule.forRoot(routes)` looks like:
```typescript
export const routes: Routes = [
    {
        path: 'lorem',
        loadChildren: 'lorem/lorem-module#LoremModule?chunkName=lorem.module'
    },
    {
        path: 'ipsum',
        loadChildren: 'ipsum/ipsum-module#IpsumModule?chunkName=ipsum.module'
    },
];
```

Terminal:
```sh
npm install webpack@beta angular-router-loader awesome-typescript-loader
node_modules/.bin/webpack
```

You end up with:
* `web/js/application.build.js`
* `web/js/lorem.module.build.js`
* `web/js/ipsum.module.build.js`

Always deliver the application.build.js with the template. The proper chunk is automatically loaded when the router hits one of the paths with `loadChildren`. Thanks to the way webpack generates the chunks you can even preload them with the initial template like `<script src="lorem.module.build.js"></script>`.

**Comment by sod** [comment]
We already moved to webpack and split the app into chunks. So the initial payload went way down. 

Regardless I tried the new view engine that landed in master today, scheduled for angular 4.0. Results are:

* total app size went down from 2.7MB to 1.7MB
* generated code went down from 1.8MB to 0.8MB
* build time (ngc + webpack) went down from 65sec. to 40sec.

Awesome! Can't wait to upgrade :)
